### ISSUE TITLE
Automatic copyright infringement detection (like ContentID)

### ISSUE THREAD

**Issue by rubenwardy** [minetest/contentdb#483]
**Note: This is just a tool to help editors find common issues.** There will be a manual override and it won't be possible for people to randomly claim textures.

## Problem

We have a lot of people uploading textures from other games. They may slightly recolor or tweak the texture, but it's still copyright infringement

## Solutions

Some code that checks whether pixel art is too similar to a source reference. It should be able to handle:

* Direct-ish matches
* Definitely recoloring
* Maybe:
	* Rotation
	* Maybe cropping / position changing (obviously, detection should have a minimum size)

I'm looking for contributions for this - if someone could write a unit-tested Python library do to this, that would be great.

I can't commit copyrighted images to this repository though. Not sure on the legalty of using a copyrighted data set to find copyright infringement

## Alternatives

This could be a third-party tool for editors to use

**Comment by JamesClarke7283** [comment]
We need to be careful with automatic Copyright Infringement detection, because YouTube does a bad job of this, and it causes a lot of problems for people.

I think its acceptable, as long as:
1. People dont get automatically blocked. (need a human in the loop to decide)
2. The content publishers, both the infringing publisher and original publisher are informed of the flag.
3. They can appeal the decision made by the human reviewer.

**Comment by ZenonSeth** [comment]
Huh this might be relevant: https://github.com/ZenonSeth/image_hash_check/

**Comment by red-001** [comment]
In my understanding contentID works by the copyright holder providing the required metadata to scan for the copyrighted work, my concern would be to what degree the pre-moderation that CDB does makes CDB a publisher not a host, even if this might have been okay in the US. 

Perhaps it would be possible to license a database containing copyrighted content hashes? In theory good actors should want to make those easy to get so their copyright doesn't get infringed in practice I think unfortunately anyone that would be litigious here is likely to be a copyright troll.

**Comment by rubenwardy** [comment]
You can see our data sets here (along with the scripts to generate them): https://github.com/ZenonSeth/image_hash_check/tree/main/hashesdb

We're storing only the perceptual hashes of the content (phash and dhash) and not the sources themselves. I believe generating these hashes is very much fair use for infringement protection

**Comment by red-001** [comment]
Yes generating the perceptual hashes sounds like textbook fair use especially in this unfortunate LLM era. It's not comprehensive however, I don't think this really raises new issues, the existing review process already might make CDB a publisher (potentially). It's really quite annoying but unfortunately sometimes trying to do the right thing and check for content before making an upload public can make that a website into a publisher not a mere host. I don't really know what solution if any exists to that other than making a best effort to keep out anything infringing and assuming copyright owners will be reasonable about it.

**Comment by ZenonSeth** [comment]
>  I don't really know what solution if any exists to that other than making a best effort to keep out anything infringing

That's kind of the reason the hash checking is being implemented, it's putting in a best-effort to keep anything infringing.

**Comment by Calinou** [comment]
> We have a lot of people uploading textures from other games. They may slightly recolor or tweak the texture, but it's still copyright infringement

For what it's worth, the [idgames archive](https://doomwiki.org/wiki/Idgames_archive) considers it acceptable to upload *modified* resources from id Software games, but not unmodified ones. Their automated checks only try to detect and [reject WADs containing unmodified id Software resources](https://www.gamers.org/pub/idgames/REJECTS).

Any kind of "fuzzy" detection is much more complex to get right, especially without access to a massive dataset of sprites from proprietary games. Not to mention the potential false positives for very small pixel art (some textures can be smaller than 16×16).

**Comment by Montandalar** [comment]
> For what it's worth, the [idgames archive](https://doomwiki.org/wiki/Idgames_archive) considers it acceptable upload _modified_ resources from id Software games, but not unmodified ones. Their automated checks only try to detect and [reject WADs containing unmodified id Software resources](https://www.gamers.org/pub/idgames/REJECTS).

That's nice, but that's id making an allowance to people who are still republishing derivative works. We operate CDB under the principle that derivative works are still infringing to avoid trouble from publishers who we do not expect to be as generous.

> Any kind of "fuzzy" detection is much more complex to get right, especially without access to a massive dataset of sprites from proprietary games. Not to mention the potential false positives for very small pixel art (some textures can be smaller than 16×16).

We can make easy gains with the main culprits that we usually get though, Minecraft base textures and texture packs, Terraria, and other proprietary large games.

**Comment by rubenwardy** [comment]
The initial goal is to detect exact (and very close) usages of commercial assets. Detecting every possible infringement shouldn't be the goal, only the most obvious. If the system has too many false positives then it loses value

**Comment by ZenonSeth** [comment]
> Any kind of "fuzzy" detection is much more complex to get right, especially without access to a massive dataset of sprites from proprietary games. Not to mention the potential false positives for very small pixel art (some textures can be smaller than 16×16).

Also wanted to comment on this because there are ways of working around this by testing how frequently textures match against things that are too vague. This is why the [hash building script](https://github.com/ZenonSeth/image_hash_check/blob/main/build_ref_hash.py) includes trimming of solid colors and various manually added paths (currently from the MC base textures)

> If the system has too many false positives then it loses value

Agreed, and the nice thing about perceptual hashing is we can tweak that threshold easily, and further tweak datasets to remove unnecessary entires.
